finality-aleph/src/data_io.rs

```diff
@@ -13,73 +13,256 @@ use futures::{
+                proposal.hash(state);
+            }
+        }
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephData<B> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (AlephData::Empty, AlephData::Empty) => true,
+            (AlephData::HeadProposal(p1), AlephData::HeadProposal(p2)) => p1.eq(p2),
+            _ => false,
         }
     }
 }
 
-pub(crate) type AlephDataFor<B> = AlephData<<B as BlockT>::Hash, NumberFor<B>>;
+impl<B: BlockT> Eq for AlephData<B> {}
+
+impl<B: BlockT> AlephData<B> {
+    pub(crate) fn is_correct(&self) -> bool {
+        match self {
+            AlephData::Empty => true,
+            AlephData::HeadProposal(proposal) => proposal.is_correct(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct AlephProposal<B: BlockT> {
+    // We expect that honest nodes create AlephData {branch: [h_0, h_1, ..., h_n], number: num} objects that represent
+    // an ascending sequence of blocks b_0, b_1, ..., b_n satisfying the following conditions:
+    //     1) hash(b_i) = h_i for i = 0, 1, ..., n,
+    //     2) parent(b_{i+1}) = b_i for i = 0, 1, ..., (n-1),
+    //     3) height(b_n) = num,
+    //     4) The parent of b_0 has been finalized (prior to creating this AlephData).
+    // Such an AlephData object should be thought of as a proposal for block b_n to be finalized.
+    pub branch: Vec<B::Hash>,
+    pub number: NumberFor<B>,
+}
+
+// Need to be implemented manually, as deriving does not work (`BlockT` is not `Hash`).
+impl<B: BlockT> Hash for AlephProposal<B> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.branch.hash(state);
+        self.number.hash(state);
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephProposal<B> {
+    fn eq(&self, other: &Self) -> bool {
+        self.number.eq(&other.number) && self.branch.eq(&other.branch)
+    }
+}
+
+impl<B: BlockT> Eq for AlephProposal<B> {}
+
+impl<B: BlockT> AlephProposal<B> {
+    pub(crate) fn new(branch: Vec<B::Hash>, block_number: NumberFor<B>) -> Self {
+        AlephProposal {
+            branch,
+            number: block_number,
+        }
+    }
+
+    pub(crate) fn is_correct(&self) -> bool {
+        if self.branch.len() > MAX_DATA_BRANCH_LEN as usize {
+            return false;
+        }
+        if self.branch.is_empty() {
+            return false;
+        }
+        if self.number < <NumberFor<B>>::saturated_from(self.branch.len()) {
+            // Honest nodes always propose a block strictly above the most recently finalized one. Also
+            // the bottom block of the branch they choose as one above the most recently finalized one.
+            return false;
+        }
+        true
+    }
+
+    pub(crate) fn top_block(&self) -> BlockHashNum<B> {
+        // Assumes that the data is correct
```

Comment:
I don't agree  -- this would generate a ton of unnecessary unwraps/expects and the only thing they do is make the lines longer and the code uglier.

There is now a single place in code which checks that a proposal is `within_bounds` and if it is not then the proposal is dropped right away. So all proposals we work with are `within_bounds` for which all these methods will always work properly. 

What I could do instead is maybe have another type, like `ValidatedProposal` (or whatever name), that is created from an `AlephProposal` as a result of the `within_bounds` check. And `ValidatedProposal` would have all these methods returning pure values and not options. 

So, either I leave it as it is, or add another type :P That's the best I can offer :)

Reply:
Perhaps even better if the appropriate variant in the `AlephData` enum was `UnvalidatedPropoposal` and `AlephProposal` would be left as it is (and would be created from the Unvalidated one).

Reply:
Yeah, this sounds good, although I would argue that we should combine this with the `within_session` check – this way we would filter out all the proposals that can be ignored regardless of the local node's knowledge about blocks.

I would also suggest that then the validated proposal can keep the hashes in an array, then there would be no need for `expect`s and the memory footprint shouldn't be problematic (the network usage will be unaffected).

Reply:
The problem is that `DataStore` is not aware of sessions (i.e., there is one `DataStore` per all sessions, and also there is currently no way of telling which session a given `Message` corresponds to). Not sure this is easy to change.

Reply:
Ok, this is done now, but the restructuring of the status checking logic is still WIP.